src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -16,10 +17,58 @@
 namespace Microsoft.Quantum.QsCompiler.DependencyAnalysis
 {
     using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
-    using TypeParameterResolutions = ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+    // Type Parameters are frequently referenced by the callable of the type parameter followed by the name of the specific type parameter.
+    using TypeParameterName = Tuple<QsQualifiedName, NonNullable<string>>;
+    using TypeParameterResolutions = ImmutableDictionary</*TypeParamterName*/ Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
```

Comment:
Should `/*TypeParamterName*/` be uncommented?

Reply:
Unfortunately C# doesn't allow previous `using` definitions to be used to define later `using` statements, so I can't actually use `TypeParameterName` in the definition of `TypeParameterResolutions` despite the fact that it should be. But I did want to put something in there so that people know that if they change `TypeParameterName`, they should also update `TypeParameterResolutions`. This is the solution I've used in the past when dealing with this issue.

Reply:
I think the types are complicated enough at this point that it would be benefit from creating a new type for `Tuple<QsQualifiedName, NonNullable<string>>`.

```csharp
struct TypeParameterName
{
    QsQualifiedName Callable { get; }
    NonNullable<string> Name { get; }

    TypeParameterName(QsQualifiedName callable, NonNullable<string> name)
    {
        this.Callable = callable;
        this.Name = name;
    }
}
```

This would also make the `typeParamName.Item1` and `typeParamName.Item2` references more readable. (Also, when C# 9 comes out it'll be easier to define immutable records like this...)

Reply:
I'm not comfortable at this time to create a new type for this. I think the typedef does an adequate job of encapsulating and communicating what these objects are. If it were possible to add tuple field names to C# typedefs, I would do that for the extra clarity of having something other than `Item1` and `Item2`, but for now the language doesn't support that.

Reply:
I think that if transformation states with one field deserve a new type, then `TypeParameterName` does too. :) If C# 9 was here, we could do

```csharp

struct TypeParameterName
{
    QsQualifiedName Callable { get; init; }
    NonNullable<string> Name { get; init; }
}
```

which does a lot to make new types feel less heavyweight, but unfortunately that isn't possible in C# 8 without making the type mutable.

Reply:
I really don't feel it's appropriate to make a new type here for this. The transformation state classes are part of the pattern that the transformations use. The typedef for TypeParameterName is a sufficient and more lightweight solution to encapsulate the tuple. I don't want the C# type checking the type parameter references throughout this code expecting a nuanced type where once the standard tuple was used to represent type parameters.

Reply:
I have never seen code that had a comment embedded inside of a type parameter before, and my first impression is that it was a mistake or typo. I agree that it would be better if C# let using aliases reference other using aliases, but since that is not possible, I think it is an indication that the type structure is too complex for aliases. Do you have any examples of similar code that uses nested aliases with embedded comments?

Also, there is a typo in the comment: `/*TypeParamterName*/` -> `/*TypeParameterName*/`

Reply:
Ah thanks for catching the typo. It's fixed now.

Reply:
Monomorphization is doing something similar with the using statements. https://github.com/microsoft/qsharp-compiler/blob/master/src/QsCompiler/Transformations/Monomorphization.cs#L18